src/semantic/type_checker.rs

```diff
@@ -0,0 +1,161 @@
+#![allow(unused_variables)]
+use lexical::types::Types;
+use syntax::expr::ExprWrapper;
+use syntax::expr::Expr::*;
+use semantic::analyzer_trait::ASTAnalyzer;
+
+pub struct TypeChecker;
+
+impl TypeChecker {
+    pub fn new() -> TypeChecker {
+        TypeChecker
+    }
+
+    fn cmp_lhs_rhs(lhs_str: String, rhs_str: String) -> Option<String> {
```

Comment:
im also confused by this. Why is this just a string vs a string? Am I being dumb in thinking that this would make the most sense as a generic, where for each type of Expr, you implement each of the match arms, rather than being generic over `ExprWrapper` in general? Then for an expression like `let a = b + c + func()`, the call heirarchy would be like (pseudocode from the deepest part of the analyzation): `oh you're analyzing func(), so look at the return type of this func based on what we know to be that function's return type ---> right now im doing plus, so my lhs: c is a ___ type, and just determined that my rhs: func() returns ___ type, assert type(rhs) == type(lhs) ---> ... ---> now im looking at equal, i know the type of rhs at this point, and i also know im letting, so this is fine, but save the type of this var now, because im declaring it`